### ISSUE TITLE
Tekton Queue. Concurrency

### ISSUE THREAD

**Issue by marniks7** [tektoncd/pipeline#5835]
### Feature request

We would like to have ability to create many PipelineRuns at once, but execute them one by one (mostly). Sometimes concurrently.

### Use case

Tekton can be used as regular pipeline \ workflow engine for all sort of activities.

### Use case#1 - Chaos Engineering
Chaos Engineering - create pipelineRuns with some chaos, e.g. deployment restart for each deployment present on the environment. Execute PipelineRun one after another. Sometimes concurrent execution maybe desirable.

### Use case#2 - Load Testing by single person

- Create pipelineRun with load equal to 10 users
- Create pipelineRun with load equal to 20 users
- Create pipelineRun with load equal to 30 users

create them at the same time, but run them one after another.

### Use case#3 - Load Testing by few people
There could be single environment for Load Testing, but multiple people working on it. In order to control the load runs done by multiple people each person can just create PipelineRun and it will be executed when previous load is finished.
Alternative: ask in chat if server is free \ available

### Solution (we have)
What we have right now:
All pipelineRuns are created in `PipelineRunPending` state. It is done manually in pipelineRun yaml. We are considering to use `Kyverno` for automatic approach, but they don't support `spec.status` field change as of September 2022.
We have implemented custom kubernetes controller which handles PipelineRuns with label `queue.tekton.dev/name` and remove `PipelineRunPending` state when previous PipelineRun finished.
```yaml
 metadata:
  labels:
    queue.tekton.dev/name: env-maria-d3hs0
```
And it is possible to search for PipelineRuns is all namespaces. E.g. in case of namespace per person
```yaml
metadata:
  annotations:
    queue.tekton.dev/scope: cluster
```

We didn't implement concurrent ability (e.g. to execute 2 PipelineRuns from the same queue), just because for top use cases we don't need it, but we may implement it in the future.

### Other Notes

- I saw this one https://github.com/tektoncd/experimental/tree/main/concurrency. I guess it could be extended to support those cases.
- I saw different tickets about concurrency around, but didn't see test cases like those.

**Comment by lbernick** [comment]
Related: https://github.com/tektoncd/pipeline/issues/4903

**Comment by khrm** [comment]
/reopen

This is part of the Roadmap and quite important.
/lifecycle frozen

**Comment by khrm** [comment]
@vdemeester Please reopen this.

**Comment by sibelius** [comment]
queueing would be awesome

**Comment by benoitschipper** [comment]
This would be a great addition, so we could pool pipeline resources from multiple customers and just use queue-ing. Pending state is not desirable as timeouts might cause the pipelines to fail on kubernetes clusters.

**Comment by sibelius** [comment]
can we do this using resource request and limits ?

**Comment by benoitschipper** [comment]
> can we do this using resource request and limits ?

Yeah you can use requests and limits for this. Problem is that if there are no resources left to run a pipeline it goes into "Pending" which eventually times out. Meaning people will get Pipeline Fails. 

If there was something like a queueing system. It would instead never timeout and just wait till there are resources available. This is obviously only for when there are busy periods on the cluster. Hence the request for something like a queueing syste.

I also found that there is a possibility to do something like a lease, but that needed additional self made resources. I want to make this a solution. For our DevOps teams.

We currently give each DevOps teams certain amount of resources to perform pipeline related tasks within a namespace. But that means that a lot of resources are potentially wasted when some DevOps teams are not running their pipelines during some days. We want to instead pool all the resources for our DevOps teams so we can reserve some nodes for pipeline related runtimes. Share the resources and utilize some sort of queueing. It's all efficiency and effectiveness related :)

Hope that makes sense :)

**Comment by sibelius** [comment]
why PENDING timeouts ?

**Comment by benoitschipper** [comment]
> why PENDING timeouts ?

Kubernetes' scheduler is, due to lack of compute space within a set quota or overall compute space on the cluster, unable to schedule the pod on any node with the requested cpu/men/storage. Making it go in pending mode.

https://kubernetes.io/docs/tasks/debug/debug-application/debug-pods/#my-pod-stays-pending

**Comment by sibelius** [comment]
can we increase PENDING timeout?

or put these tasks in a "queue"?

like the error CouldNotGetTask

**Comment by benoitschipper** [comment]
Here I found the feature request:
https://github.com/tektoncd/pipeline/issues/5835

**Comment by gbenhaim** [comment]
/remove-lifecycle frozen
Controlling the concurrency amount is a critical feature of a pipeline framework. Today it's possible to achieve that using Kueue and the tekton-kueue plug-in, which I'm going to propose to the community.

https://github.com/konflux-ci/tekton-kueue

**Comment by sibelius** [comment]
how to use this https://github.com/konflux-ci/tekton-kueue ?
